### PR TITLE
Overhaul of Region Management UI

### LINKED ISSUE #548: Improve Management and Display of Regions and Custom API Servers
### Expected behavior

OBA should make it more clear than it does when you're connected to a custom API server. It should tell which one you're connected to—i.e. the URL and any other information it might have.

### Actual behavior

It's not super-clear about this right now.

Furthermore, the Regions management view controller is just generally problematic. It'd be great to rebuild it on top of OBAStaticTableViewController.

Related to #305

### LINKED ISSUE #677: Reduce likelihood of crashes in generating feedback messages
The method `-[OBAEmailHelper messageBodyForModelDAO:currentLocation:]` consumes a bundled HTML file called `feedback_message_body.html`. This HTML file contains several `%@` string formatting replacement points. 

Using these `%@` replacement values is incredibly brittle, and prone to awful, obscure crashes. It would be far better to replace them with a Mustache-like syntax, like `{{app_version}}`, which can be replaced with the far safer NSString method `-stringByReplacingOccurrencesOfString:withString:`.

Steps to successfully closing this issue:

- [ ] Replace the `%@` values in `feedback_message_body.html` with Mustache-like `{{variable_name}}` values
- [ ] Replace the Mustache-like values with actual values in `OBAEmailHelper`
- [ ] Write a test or two validating that the new system works properly (n.b. this may require some refactoring. Feel free to ping me about how I think this should be done.)

### PR DESCRIPTION
**Pull request by aaronbrethorst**
Major Features:

* Rebuild the Region List View Controller (and do it in Swift)
* Rebuild the New Custom API view controller (now called the Region Builder)

Fixes:

* Fixes #548: Improve Management and Display of Regions and Custom API Servers
* Fixes #677: Reduce likelihood of crashes in generating feedback messages

Region List Improvements:

* Rebuild the Region List/Picker controller on top of the static table system and do it in Swift for good measure. This controller has far less code in it and perhaps a couple less features, but definitely far fewer bugs.
* Automatically set the current region to be the most-recently-created custom region

Custom API Server:

So, it used to be that custom APIs were set by the user who would type in a URL that pointed to their custom OBA server. This is still the case, however now it has been greatly expanded. You no longer type in *just* your server URL, but now specify the entire region. There are, I think, still some bugs in the codebase around custom API server management, but multi-region support is quite solid. By bypassing the grafted-on custom API server stuff and sticking strictly to the multi-region code, things should be far smoother.

Static Table Editor Mode:

* Create a text field row and cell combo for the static table system
* Make it possible to use the static table system as an editor by dumping the user's inputted values into an NSMutableDictionary
* Keyboard management for static table controller: the keyboard's appearance and disappearance will cause the scroll and content insets of the table to automatically update

Static Table:

* Fix bugs in OBASwitchCell
* OBATableRows support selection style
* Add a standardized delete method to the OBAStaticTableViewController: -deleteRowAtIndexPath:
* Add out of bounds safeguards to -[OBAStaticTableViewController rowAtIndexPath:]
* Add a standardized remove method to OBABaseSection

General Regions Work:

* Replace the confusingly-named readSetRegionAutomatically and writeSetRegionAutomatically methods on OBAModelDAO with a property named `automaticallySelectRegion`
* Add an -isEqual: method to OBARegionV2
* Change confusing method names in OBAApplicationDelegate
* Move OBARegionHelper into OBAKit, and make necessary modifications to make this work properly

Misc:

* Move the Swift bridging header to a more reasonable location in the project
* Add more GPX files to project
* Move common localizable strings into a new class called OBAStrings
* rename OBAModelDAO.region to currentRegion
* Fix a warning saying that the map view was attempting to access the user's location before receiving authorization.
* improve feedback email construction process. Do a string replacement instead of a formatting string-oriented approach.

### CODE REVIEW

File: OBAApplicationDelegate.h [hunk 1]
```diff
@@ -15,31 +15,27 @@
...
 @class OBARecentStopsViewController;
 @class OBAInfoViewController;
-@class OBARegionListViewController;
 
 @interface OBAApplicationDelegate : UIResponder <UIApplicationDelegate>
-@property(nonatomic, strong) UIWindow *window;
+/**
+ Abstracts OBAModelDAO setters / getters and calls the appropriate analytics methods.
+ */
+@property(nonatomic,assign) BOOL automaticallySelectRegion;
```
Comment (aaronbrethorst) [inline OBAApplicationDelegate.h]: see if this can be removed.

File: OBAKit/Helpers/OBAStrings.m [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+//  Created by Aaron Brethorst on 8/25/16.
+//  Copyright © 2016 OneBusAway. All rights reserved.
+//
+
+#import "OBAStrings.h"
+
+@implementation OBAStrings
+
++ (NSString*)cancel {
+    return NSLocalizedString(@"Cancel", @"");
```
Comment (aaronbrethorst) [inline OBAKit/Helpers/OBAStrings.m]: add comments for each of these strings

File: OBAKit/Location/OBARegionHelper.m [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+- (OBAListWithRangeAndReferencesV2*)loadDefaultRegions {
+
+    NSLog(@"Unable to retrieve regions file. Loading default regions from the app bundle.");
+
+    OBAModelFactory *factory = [OBAApplication sharedApplication].modelService.modelFactory;
+    NSError *error = nil;
+
+    NSData *data = [[NSData alloc] initWithContentsOfFile:[[NSBundle mainBundle] pathForResource:@"regions-v3" ofType:@"json"]];
+
+    if (!data) {
```
Comment (aaronbrethorst) [inline OBAKit/Location/OBARegionHelper.m]: `OBAGuard`

File: OBAKit/Location/OBARegionHelper.m [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    OBAListWithRangeAndReferencesV2 *references = [factory getRegionsV2FromJson:defaultJSONData error:&error];
+
+    if (error) {
+        NSLog(@"Issue parsing bundled JSON data: %@", error);
+    }
+
+    return references;
+}
+
+- (void)processRegionData:(id)regionData {
```
Comment (aaronbrethorst) [inline OBAKit/Location/OBARegionHelper.m]: change parameter to be `OBAListWithRangeAndReferencesV2`

File: OBAKit/Location/OBARegionHelper.m [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+            return NSOrderedAscending;
+        }
+        else {
+            return NSOrderedSame;
+        }
+     }];
+
+    NSString *oldRegion = @"null";
+
+    if ([OBAApplication sharedApplication].modelDao.currentRegion != nil) {
```
Comment (aaronbrethorst) [inline OBAKit/Location/OBARegionHelper.m]: remove `!= nil`

File: OBAKit/Location/OBARegionHelper.m [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+        oldRegion = [OBAApplication sharedApplication].modelDao.currentRegion.regionName;
+    }
+
+    [OBAApplication sharedApplication].modelDao.currentRegion = self.regions[0];
+    [[OBAApplication sharedApplication] refreshSettings];
+    [[OBAApplication sharedApplication].locationManager removeDelegate:self];
+    [OBAApplication sharedApplication].modelDao.automaticallySelectRegion = YES;
+
+    [self.delegate regionHelper:self automaticallyConfiguredRegionWithName:[OBAApplication sharedApplication].modelDao.currentRegion.regionName oldRegion:oldRegion];
```
Comment (aaronbrethorst) [inline OBAKit/Location/OBARegionHelper.m]: not useful, remove.

File: OBAKit/Location/OBARegionHelper.m [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+    [[NSNotificationCenter defaultCenter] postNotificationName:OBARegionDidUpdateNotification object:nil];
+}
+
+- (void)setRegion {
+    NSString *regionName = [OBAApplication sharedApplication].modelDao.currentRegion.regionName;
+
+    if (regionName) {
```
Comment (aaronbrethorst) [inline OBAKit/Location/OBARegionHelper.m]: invert and return immediately

File: OBAKit/Models/dao/OBAModelDAO.m [hunk 8]
```diff
@@ -483,53 +501,8 @@ - (void)setVisited:(BOOL)visited forSituationWithId:(NSString*)situationId {
...
-
-    NSInteger over = [_mostRecentCustomApiUrls count] - kMaxEntriesInMostRecentList;
-    for (NSInteger i=0; i<over; i++) {
-        [_mostRecentCustomApiUrls removeObjectAtIndex:_mostRecentCustomApiUrls.count - 1];
-    }
-
-    [_preferencesDao writeMostRecentCustomApiUrls:_mostRecentCustomApiUrls];
-}
-
 - (NSString*)normalizedAPIServerURL {
```
Comment (aaronbrethorst) [inline OBAKit/Models/dao/OBAModelDAO.m]: check to see how easy it would be to replace this with something that would just return an NSURL

**Comment by aaronbrethorst** [comment]
Other TODOs:

- [ ] Test and improve error cases for regions. e.g. If the custom region has an incorrect URL for downloading agencies, show the user what the attempted URL was.

File: OBAKit/Services/OBAModelFactory.m [hunk 9]
```diff
@@ -438,8 +445,17 @@ - (void) addTripDetailsV2RulesWithPrefix:(NSString*)prefix {
 
 - (void) addAgencyWithCoverageV2RulesWithPrefix:(NSString*)prefix {
     [self addObjectCreateRule:[OBAAgencyWithCoverageV2 class] forPrefix:prefix];
-    [self addSetPropertyRule:@"agencyId" forPrefix:[self extendPrefix:prefix withValue:@"agencyId"]];    
+    [self addSetPropertyRule:@"agencyId" forPrefix:[self extendPrefix:prefix withValue:@"agencyId"]];
     [self addSetCoordinatePropertyRule:@"coordinate" withPrefix:prefix method:OBASetCoordinatePropertyMethodLatLon];
+
+    //
```
Comment (aaronbrethorst) [inline OBAKit/Services/OBAModelFactory.m]: delete unnecessary `//`s

File: OBAKit/Services/OBARegionValidator.h [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+@interface OBARegionValidator : NSObject
+@property(nonatomic,copy,nullable) NSString *APIServerURLString;
+@property(nonatomic,copy) NSURL *APIServerURL;
+
+- (instancetype)initWithAPIServerURLString:(NSString*)URLString NS_DESIGNATED_INITIALIZER;
+- (instancetype)init NS_UNAVAILABLE;
+
+/**
+ Validates that API calls can successfully be made to the specified OBA API Server
```
Comment (aaronbrethorst) [inline OBAKit/Services/OBARegionValidator.h]: this comment is incorrect.

File: OBAKit/Services/OBARegionValidator.h [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+//  Created by Aaron Brethorst on 8/25/16.
+//  Copyright © 2016 OneBusAway. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+#import <PromiseKit/PromiseKit.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface OBARegionValidator : NSObject
```
Comment (aaronbrethorst) [inline OBAKit/Services/OBARegionValidator.h]: I bet I can come up with a better class name than this. It was originally semi-relevant but the purpose of the class changed.

File: OBAKit/Services/OBARegionValidator.m [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+@end
+
+@implementation OBARegionValidator
+@dynamic APIServerURLString;
+
+- (instancetype)initWithAPIServerURLString:(NSString*)URLString {
+    self = [super init];
+
+    if (self) {
+        // normally you should directly access the ivars when
```
Comment (aaronbrethorst) [inline OBAKit/Services/OBARegionValidator.m]: still true?

File: Resources/feedback_message_body.html [hunk 13]
```diff
@@ -14,8 +14,6 @@
 
 <p>Region: %@</p>
 
-<p>Custom API: %@</p>
```
Comment (aaronbrethorst) [inline Resources/feedback_message_body.html]: be sure to test this and make sure it doesn't break!

File: ui/info/OBAInfoViewController.m [hunk 14]
```diff
@@ -52,6 +53,14 @@ - (void)viewWillAppear:(BOOL)animated {
     [self reloadData];
 }
 
+- (void)viewDidAppear:(BOOL)animated {
```
Comment (aaronbrethorst) [inline ui/info/OBAInfoViewController.m]: delete me!

File: ui/regions/RegionBuilderViewController.swift [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+    }()
+
+    lazy var nameRow: OBATextFieldRow = {
+        let row = OBATextFieldRow.init(labelText: NSLocalizedString("Name", comment: ""), textFieldText: self.region.regionName)
+        RegionBuilderViewController.applyPropertiesToTextRow(row, model: self.userDataModel)
+        return row
+    }()
+
+    lazy var siriRealTimeRow: OBASwitchRow = {
+        let row = OBASwitchRow.init(title: "Supports Siri Realtime APIs", action: nil)
```
Comment (aaronbrethorst) [inline ui/regions/RegionBuilderViewController.swift]: NSLocalizedString

File: ui/regions/RegionBuilderViewController.swift [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+
+import UIKit
+import OBAKit
+import SVProgressHUD
+
+@objc protocol RegionBuilderDelegate {
+    func regionBuilderDidCreateRegion(region: OBARegionV2)
+}
+
+class RegionBuilderViewController: OBAStaticTableViewController {
```
Comment (aaronbrethorst) [inline ui/regions/RegionBuilderViewController.swift]: A ton of the rows that are displayed here never actually get used. Hide or delete them.

File: ui/regions/RegionBuilderViewController.swift [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+        let row = OBATextFieldRow.init(labelText: NSLocalizedString("Contact Email", comment: ""), textFieldText: self.region.contactEmail)
+        row.keyboardType = .EmailAddress
+        row.autocapitalizationType = .None
+        RegionBuilderViewController.applyPropertiesToTextRow(row, model: self.userDataModel)
+        return row
+    }()
+
+    lazy var twitterURLRow: OBATextFieldRow = {
+        let row = OBATextFieldRow.init(labelText: NSLocalizedString("Twitter URL", comment: ""), textFieldText: self.region.twitterUrl)
+        row.keyboardType = .URL
```
Comment (aaronbrethorst) [inline ui/regions/RegionBuilderViewController.swift]: wrong keyboard type

File: ui/regions/RegionBuilderViewController.swift [hunk 18]
```diff
@@ -0,0 +1,256 @@
...
+        self.navigationItem.rightBarButtonItem = UIBarButtonItem.init(title: OBAStrings.save(), style: .Done, target: self, action: #selector(save))
+
+        let requiredSection = OBATableSection.init(title: NSLocalizedString("Required", comment: ""))
+
+        requiredSection.addRow(self.baseURLRow)
+        requiredSection.addRow(self.nameRow)
+
+        let optionalSection = OBATableSection.init(title: NSLocalizedString("Optional", comment: ""))
+
+        optionalSection.addRow(self.siriBaseURLRow)
```
Comment (aaronbrethorst) [inline ui/regions/RegionBuilderViewController.swift]: i don't like this order

File: ui/regions/RegionBuilderViewController.swift [hunk 19]
```diff
@@ -0,0 +1,256 @@
...
+
+            if let delegate = self.delegate {
+                delegate.regionBuilderDidCreateRegion(self.region)
+            }
+
+            self.dismissViewControllerAnimated(true, completion: nil)
+        }.always {
+            SVProgressHUD.dismiss()
+        }.error { error in
+            let msg = String(format: NSLocalizedString("Please check the URL and try again.\r\n\r\n%@", comment: ""), (error as NSError).localizedDescription)
```
Comment (aaronbrethorst) [inline ui/regions/RegionBuilderViewController.swift]: Base this on top of the new AlertPresenter class.

File: ui/regions/RegionListViewController.swift [hunk 20]
```diff
@@ -0,0 +1,233 @@
...
+        }.then { location in
+            self.currentLocation = location as? CLLocation
+            return OBAApplication.sharedApplication().modelService.requestRegions()
+        }.then { regions -> Void in
+            self.regions = regions as? [OBARegionV2]
+            self.loadData()
+        }.always {
+            SVProgressHUD.dismiss()
+        }.error { error in
+            // TODO: FIXME
```
Comment (aaronbrethorst) [inline ui/regions/RegionListViewController.swift]: FIXME! :)

File: ui/static_table/OBAStaticTableViewController.m [hunk 21]
```diff
@@ -136,6 +155,21 @@ - (void)replaceRowAtIndexPath:(NSIndexPath*)indexPath withRow:(OBABaseRow*)row {
     section.rows = [NSArray arrayWithArray:rows];
 }
 
+- (void)deleteRowAtIndexPath:(NSIndexPath*)indexPath {
```
Comment (aaronbrethorst) [inline ui/static_table/OBAStaticTableViewController.m]: look into ease of adopting this on OBABookmarkViewController.

**Comment by barbeau** [comment]
@aaronbrethorst IIRC OBA iOS hits the Regions API on every startup.  This really isn't necessary, and we could get away with far less frequent updates, which should help with both device and server performance.  OBA Android refreshes weekly.  If you're overhauling the region-related code you might want to consider this change.  This also helps us roll out features like trip planning (adding an OTP URL to a region) in a little more gradual manner, rather than flipping the switch and immediately going live in production for all users.  It's a crude method for controlling a rollout server-side, but it seems to work ok.

**Comment by aaronbrethorst** [comment]
@barbeau I have opened #678 to address this for the upcoming release.

**Comment by barbeau** [comment]
Awesome, thanks!

**Pull request merged**